### ISSUE TITLE
Verify that we get site isolation in Tor

### ISSUE THREAD

**Issue by riastradh-brave** [brave/browser-laptop#14390]
During some usability tests we seemed to get the same IP address more often than expected.  Figure out why and whether there is a bug here.

**Comment by diracdeltas** [comment]
Based on my testing I'm pretty sure there is a bug here.

STR:
1. search 'what is my IP' on a tor tab
2. remember what duckduckgo says the IP is
3. click on a search result that shows you your IP. 
4. note it's always the same as what duckduckgo said even though the site is a different origin

**Comment by darkdh** [comment]
And I fount out doing a **Refresh** after step 3, it show different ip than DDG.
So it seems like proxy setting is not able to applied in time before the request processed

**Comment by riastradh-brave** [comment]
Examination on the wire between brave and the tor daemon reveals:

1. a socks request with username duckduckgo.com to host duckduckgo.com when you do the search
2. a socks request with username iplocation.net to host iplocation.net when you open the link in a new tab
3. a socks request with username duckduckgo.com to host iplocation.net when you click on a link in the same tab
4. a socks request with username iplocation.net when you reload the page

**Comment by darkdh** [comment]
I'm working on a fix which will block request until proxy config is reload

**Comment by riastradh-brave** [comment]
Is this a race between the caller of CreateRequestContextForStoragePartition and TorSetProxy or something?  Why doesn't the CONFIG_UNSET return from GetLatestProxyConfig already block it?

**Comment by kjozwiak** [comment]
@diracdeltas mind adding a test case for QA?

**Comment by diracdeltas** [comment]
@kjozwiak yup, the test plan is basically https://github.com/brave/browser-laptop/issues/14390#issuecomment-396682930t - i'll copy it into the issue

**Comment by kjozwiak** [comment]
Thanks @diracdeltas 👍 Wasn't sure if those were sufficient enough so figured I would ask you just incase!

**Comment by srirambv** [comment]
Verified on Windows 10 x64 using
* 0.23.16 - `2042c45`
* Muon - `7.1.1`
* libchromiumcontent - `67.0.3396.87`
